Fix missing depends_on for py-torch dependency in cp2k

mtaillefumier opened the pull request:
fix for issue https://github.com/spack/spack/issues/41445, #41521 (thanks @yizeyi18)

- QUIP needs far more work because they use a custom build system. Transition to cmake or other build systems not started yet. 
- fix fpic libvori issue -> PR #41631
- enforce propagation of cuda and amd architecture to all packages -> #41521

Commit:
Fix missing depends_on for py-torch dependency

RMeli commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -968,6 +973,8 @@ def cmake_args(self):
                     self.define("CP2K_WITH_GPU", gpu_ver),
                 ]
 
+        # we need to add the quip package to spack.
+
```
I think this would be better off as an issue (in Spack), instead of a comment here? I'd also remove the related code, instead of commenting it out. I don't know what the best practice is in Spack.

mtaillefumier replied:
Agreed. I removed the comment and will open an issue (eventually).

yizeyi18 commented:
py-torch installed from spack seems contain no FindTorch.cmake. Would cmake really find pytorch with CMAE_PREFIX_PATH?
![image](https://github.com/spack/spack/assets/119801373/e3d1c7f5-ee38-447e-a998-f97b30392965)

mtaillefumier commented:
> py-torch installed from spack seems provide no FindTorch.cmake. Would cmake really find pytorch with only CMAKE_PREFIX_PATH? ![image](https://private-user-images.githubusercontent.com/119801373/289063346-e3d1c7f5-ee38-447e-a998-f97b30392965.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIyOTA2MDQsIm5iZiI6MTcwMjI5MDMwNCwicGF0aCI6Ii8xMTk4MDEzNzMvMjg5MDYzMzQ2LWUzZDFjN2Y1LWVlMzgtNDQ3ZS1hOTk4LWY5N2IzMDM5Mjk2NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIxMVQxMDI1MDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MmQ1YTM2MmRjNmQ3OGE5M2QwYTIxZjZjNGNjYWY0OTQyN2ZkYjY2ZDI1NTY1MGQyMDMxZTkwYzJhYjE5ZjUyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.mkvSAl6LbSD43EGEEP1GR3sSm4FuZc6PrPCr4qjtzEw)

the problem is that py-torch does not install the cmake files at the right location but spack should be aware of this. And yes CMAKE_PREFIX_PATH is the right variable to set up in that case but it is not cp2k job to know this. If it fails to find the related configurations files then the problem should be reported to the py-torch maintainers not cp2k.

Commit:
Add missing rocm arch in the cmake section

Signed-off-by: Mathieu Taillefumier <mathieu.taillefumier@free.fr>

mtaillefumier commented:
I reply to my own question. py-torch does properly install the cmake files. spack does its job and set the CMAKE_PREFIX=PATh properly.

yizeyi18 commented on var/spack/repos/builtin/packages/libvori/package.py:
```diff
@@ -19,3 +19,11 @@ class Libvori(CMakePackage):
+
+class CMakeBuilder(spack.build_systems.cmake.CMakeBuilder):
+    def cmake_args(self):
+        args = [self.define_from_variant("CMAKE_POSITION_INDEPENDENT_CODE", "fpic")]
```
There are spack keys like “self.compiler.cxx_pic_flag”, "self.compiler.cc_pic_flag" or "self.compiler.f77_pic_flag" to set PIC compiler flag; see superlu or other packages provide the pic variant.

mtaillefumier replied:
I am not setting the compiler flag explicitly. CMake does it for me as it should be. If I am not mistaken `self.compiler:.....` is used when we need to generate makefiles by hand which is not our case here.

yizeyi18 commented:
>update : could you give the output of the `export` command after loading the pytorch package.
> 
> ```
> spack load py-torch
> export | grep "CMAKE" 
> ```

The output is as below:
<details><summary>output</summary>
<pre>
declare -x CMAKE_PREFIX_PATH="/opt/intel/oneapi/mkl/2024.0/lib/cmake::/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-torch-2.1.1-jpzzlydywwa6rq7avtokcpgjoleiogvt:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/valgrind-3.20.0-vnufikj7mhbpe2kpsu7vcez47h5mu7d6:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/boost-1.82.0-d5ul7ybylxowmhvfvzvj2gsyecwbm2le:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/zstd-1.5.5-y3b4mi5nmxbizogtj4qv4hszkyx524yc:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/sleef-3.5.1_2020-12-22-yc2migga4vcclirlizgx4awdciqdtrtk:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-typing-extensions-4.8.0-hxs3lgafen2w5tc5tpypfvc54o6eucje:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-tqdm-4.66.1-wcsvpebsj4xncooagxotw76sztpgrmyf:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-sympy-1.11.1-6dktwcmotlnbt2ln3klml46qqm722lu5:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-mpmath-1.2.1-fhwonzs6tye72x73tmvi5oe2b5jehu3p:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-requests-2.31.0-qz7grk2unxmqnt4nhpqdcsqqtocbg6ps:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-urllib3-2.1.0-hzv3ivdofna2nyy2qnqr6f3zqoffnpkv:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-idna-3.4-35lxbors2v2viy63w766a7r3bwrgsobe:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-charset-normalizer-3.3.0-ab2mpxsmoeqp5fazthva3up7j5mzsxcc:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-certifi-2023.7.22-4wivpbdfra7fby2ogr6hryrw4pgk3pda:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-pyyaml-6.0-qfwjajuayhnj7a7zggc2d2rpfpodpyso:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libyaml-0.2.5-z4uizohecp3gbd7xwl6ynfjmi6pybq7j:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-pybind11-2.11.0-ojjly3sc6dfe3k4hmxjep2bnwz5uc53z:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-protobuf-3.20.3-fazy4yt6q4lkhbetntyn4kbfzktvviz7:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-setuptools-68.0.0-qqubkhng5e2ekust6xcrqha6uueeawvf:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-numpy-1.26.2-qp3rdkkp4hx5shewrojz6kopucgm3ewk:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-networkx-2.7.1-z46nrs23gaxvc3tkewr2vee6y4pbivbb:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-jinja2-3.1.2-tqgyuct7kastkyovute3t6q27uqj245p:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-markupsafe-2.1.3-cimnzy4w7fmotd62vmgjl3lg22yssqf2:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-fsspec-2023.10.0-rwkhpg52hfklejtqbh3tjjhlxxp26trm:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-filelock-3.12.4-azaa3rio3xz7a7qhgkwqjsojunxzi6jp:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-astunparse-1.6.3-lnm2huvqlponzebstzognqklxyb53t74:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-wheel-0.37.1-ylqhqpwsxq6myyr5dhgmtliidxtvfwsr:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-six-1.16.0-2nwkccb7fht3kjfzaufqhhl46a4qkspj:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/python-3.10.12-dz6f2fukkjyda3lfi52p3rop6ufrztjl:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/util-linux-uuid-2.38.1-rfhgv3uuhv45dq7fm2aqendfostdmcbq:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/sqlite-3.42.0-5gtzh6fnsdomjf6zj3qq6b4f6dmwcldn:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libxcrypt-4.4.35-ns5xr76ky62xhjihmtcg4rpcolislnsu:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libffi-3.4.4-xfezfptvmnqub44nedtus6fnqibg4wve:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/gettext-0.21.1-yxbg4m4nysxkcf5yka4yuzo3swa7622g:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/gdbm-1.23-gzn6aaio4pihy5ifz2catqk4ejk6gwmg:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/readline-8.2-7lgwtuy2zjbgf6cd5fvhx57kkm5zetgq:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/expat-2.5.0-mjzoqpkt2qrbcloyxhclqhx4mpzs7w4v:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libbsd-0.11.7-5cus6qoskqmzqjaei3luvxsdm2tdec7j:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libmd-1.0.4-6fz5v2fkdmhict5rtg4rt7qbzu7aohzb:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/bzip2-1.0.8-exxvknagepfzwv7ctxv5eyobxmy5nchv:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/pthreadpool-2021-04-13-353l43cdj3v533eynsozqj4us37pychf:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/psimd-2020-05-17-h4rn6ojgyqtaiii5xjtpgamkfwrawr22:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/protobuf-3.24.3-dopke7is4mupyduloswubie4sbr2kfyl:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/abseil-cpp-20230125.3-i6pkxk53gkvwoybthypc5f5267nnvk3e:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/openmpi-4.1.5-wtajzbikckprkfytzm7eakiebknu6bkf:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/pmix-4.2.4-hxphdi7alv3a4on2aw2lt44oq6hxvveu:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libevent-2.1.12-gpazfhhgigj4kiyehtl22kbfa7kxxq55:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/openssl-3.1.1-hc7n5fodkmxq45v5wrkqnkngyvwtzdyz:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/hwloc-2.9.1-mboh5p42o4mumeq7mnxky42xq457ehbi:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/ncurses-6.4-es2yxb44oxza24zzozv7msiiv4yybl7y:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libxml2-2.10.3-x62zjudsnalunonmuxzcdgvvalnumzsl:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/zlib-1.2.13-vtmsxkgzqja6qtrafmi54dysycyaxixv:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/xz-5.4.1-snzvo3ymih6mcb4gknr736xtyfpoo6t3:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libiconv-1.17-ked53pc6rkl446jk2t2rpfjtsrbogbfh:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libpciaccess-0.17-aovdawcrczq7dvr4y4x5fs7lmvbpargh:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/openblas-0.3.23-2fnivlg4e7pdfxtdv5oxs4cr3fgao3r6:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/numactl-2.0.14-mkfs2bwy36bfp7qq6cvhkpyn5vxk7vk5:/opt/intel/oneapi:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/gloo-2023-05-19-4bareiglrysulwi2vn4uwuwaa47f5ppc:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/fxdiv-2020-04-17-bnny2pwzfpqstnhoar4hxpwmfivkyjpy:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/fp16-2020-05-14-5fyfb5tfqwqbjg67pxlbv3sjaecm7zvi:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/eigen-3.4.0-qqdmuoqto3o43thcj6osst65r7g5nnzo"
</pre></details>


Now I know that cmake would find py-torch properly. A new problem is that my build attempt failed with an error reporting "'is_same_v' is not a member of 'std'", referred in .../torch/include/ATen/core/boxing/impl/boxing.h; it seems this is part of c++17, but the compile flags have --std=gnu++11. I haven's find where introduced this flag yet; had you faces issue similar?

mtaillefumier commented:
> > update : could you give the output of the `export` command after loading the pytorch package.
> > ```
> > spack load py-torch
> > export | grep "CMAKE" 
> > ```
> 
> The output is as below:
> output
> 
> declare -x CMAKE_PREFIX_PATH="/opt/intel/oneapi/mkl/2024.0/lib/cmake::/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-torch-2.1.1-jpzzlydywwa6rq7avtokcpgjoleiogvt:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/valgrind-3.20.0-vnufikj7mhbpe2kpsu7vcez47h5mu7d6:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/boost-1.82.0-d5ul7ybylxowmhvfvzvj2gsyecwbm2le:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/zstd-1.5.5-y3b4mi5nmxbizogtj4qv4hszkyx524yc:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/sleef-3.5.1_2020-12-22-yc2migga4vcclirlizgx4awdciqdtrtk:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-typing-extensions-4.8.0-hxs3lgafen2w5tc5tpypfvc54o6eucje:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-tqdm-4.66.1-wcsvpebsj4xncooagxotw76sztpgrmyf:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-sympy-1.11.1-6dktwcmotlnbt2ln3klml46qqm722lu5:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-mpmath-1.2.1-fhwonzs6tye72x73tmvi5oe2b5jehu3p:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-requests-2.31.0-qz7grk2unxmqnt4nhpqdcsqqtocbg6ps:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-urllib3-2.1.0-hzv3ivdofna2nyy2qnqr6f3zqoffnpkv:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-idna-3.4-35lxbors2v2viy63w766a7r3bwrgsobe:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-charset-normalizer-3.3.0-ab2mpxsmoeqp5fazthva3up7j5mzsxcc:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-certifi-2023.7.22-4wivpbdfra7fby2ogr6hryrw4pgk3pda:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-pyyaml-6.0-qfwjajuayhnj7a7zggc2d2rpfpodpyso:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libyaml-0.2.5-z4uizohecp3gbd7xwl6ynfjmi6pybq7j:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-pybind11-2.11.0-ojjly3sc6dfe3k4hmxjep2bnwz5uc53z:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-protobuf-3.20.3-fazy4yt6q4lkhbetntyn4kbfzktvviz7:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-setuptools-68.0.0-qqubkhng5e2ekust6xcrqha6uueeawvf:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-numpy-1.26.2-qp3rdkkp4hx5shewrojz6kopucgm3ewk:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-networkx-2.7.1-z46nrs23gaxvc3tkewr2vee6y4pbivbb:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-jinja2-3.1.2-tqgyuct7kastkyovute3t6q27uqj245p:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-markupsafe-2.1.3-cimnzy4w7fmotd62vmgjl3lg22yssqf2:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-fsspec-2023.10.0-rwkhpg52hfklejtqbh3tjjhlxxp26trm:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-filelock-3.12.4-azaa3rio3xz7a7qhgkwqjsojunxzi6jp:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-astunparse-1.6.3-lnm2huvqlponzebstzognqklxyb53t74:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-wheel-0.37.1-ylqhqpwsxq6myyr5dhgmtliidxtvfwsr:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/py-six-1.16.0-2nwkccb7fht3kjfzaufqhhl46a4qkspj:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/python-3.10.12-dz6f2fukkjyda3lfi52p3rop6ufrztjl:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/util-linux-uuid-2.38.1-rfhgv3uuhv45dq7fm2aqendfostdmcbq:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/sqlite-3.42.0-5gtzh6fnsdomjf6zj3qq6b4f6dmwcldn:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libxcrypt-4.4.35-ns5xr76ky62xhjihmtcg4rpcolislnsu:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libffi-3.4.4-xfezfptvmnqub44nedtus6fnqibg4wve:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/gettext-0.21.1-yxbg4m4nysxkcf5yka4yuzo3swa7622g:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/gdbm-1.23-gzn6aaio4pihy5ifz2catqk4ejk6gwmg:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/readline-8.2-7lgwtuy2zjbgf6cd5fvhx57kkm5zetgq:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/expat-2.5.0-mjzoqpkt2qrbcloyxhclqhx4mpzs7w4v:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libbsd-0.11.7-5cus6qoskqmzqjaei3luvxsdm2tdec7j:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libmd-1.0.4-6fz5v2fkdmhict5rtg4rt7qbzu7aohzb:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/bzip2-1.0.8-exxvknagepfzwv7ctxv5eyobxmy5nchv:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/pthreadpool-2021-04-13-353l43cdj3v533eynsozqj4us37pychf:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/psimd-2020-05-17-h4rn6ojgyqtaiii5xjtpgamkfwrawr22:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/protobuf-3.24.3-dopke7is4mupyduloswubie4sbr2kfyl:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/abseil-cpp-20230125.3-i6pkxk53gkvwoybthypc5f5267nnvk3e:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/openmpi-4.1.5-wtajzbikckprkfytzm7eakiebknu6bkf:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/pmix-4.2.4-hxphdi7alv3a4on2aw2lt44oq6hxvveu:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libevent-2.1.12-gpazfhhgigj4kiyehtl22kbfa7kxxq55:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/openssl-3.1.1-hc7n5fodkmxq45v5wrkqnkngyvwtzdyz:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/hwloc-2.9.1-mboh5p42o4mumeq7mnxky42xq457ehbi:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/ncurses-6.4-es2yxb44oxza24zzozv7msiiv4yybl7y:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libxml2-2.10.3-x62zjudsnalunonmuxzcdgvvalnumzsl:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/zlib-1.2.13-vtmsxkgzqja6qtrafmi54dysycyaxixv:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/xz-5.4.1-snzvo3ymih6mcb4gknr736xtyfpoo6t3:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libiconv-1.17-ked53pc6rkl446jk2t2rpfjtsrbogbfh:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/libpciaccess-0.17-aovdawcrczq7dvr4y4x5fs7lmvbpargh:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/openblas-0.3.23-2fnivlg4e7pdfxtdv5oxs4cr3fgao3r6:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/numactl-2.0.14-mkfs2bwy36bfp7qq6cvhkpyn5vxk7vk5:/opt/intel/oneapi:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/gloo-2023-05-19-4bareiglrysulwi2vn4uwuwaa47f5ppc:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/fxdiv-2020-04-17-bnny2pwzfpqstnhoar4hxpwmfivkyjpy:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/fp16-2020-05-14-5fyfb5tfqwqbjg67pxlbv3sjaecm7zvi:/home/yizeyi18/soft/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-13.1.0/eigen-3.4.0-qqdmuoqto3o43thcj6osst65r7g5nnzo"
> 
> Now I know that cmake would find py-torch properly. A new problem is that my build attempt failed with an error reporting "'is_same_v' is not a member of 'std'", referred in .../torch/include/ATen/core/boxing/impl/boxing.h; it seems this is part of c++17, but the compile flags have --std=gnu++11. I had not found where introduced this flag yet; had you faced issue similar?

I had similar output on my local machine for py-torch.... the cmake files are not located where they should be. 

compilation error : Hmm, never faced this issue but I should try. I can always turn on c++17 by default though

mtaillefumier commented:
also note these errors have nothing to do with the PR itself.

Commit:
force gpu architecture constrain to package supporting gpu

Signed-off-by: Mathieu Taillefumier <mathieu.taillefumier@free.fr>

Commit:
Various updates

- Updated Mi300 architecture info (need cp2k master)
- remove comment about quip (will open an issue

mtaillefumier commented:
the pic issue is solved in a separate PR.

Commit:
fix typo in the variant name

Commit:
Fix faulty strings dependencies

msimberg commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -109,7 +109,7 @@ class Cp2k(MakefilePackage, CudaPackage, CMakePackage, ROCmPackage):
     )
     variant("pytorch", default=False, description="Enable libtorch support")
-    variant("quip", default=False, description="Enable quip support")
+    # variant("quip", default=False, description="Enable quip support")
```
Remove completely?

msimberg commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -351,11 +370,27 @@ class Cp2k(MakefilePackage, CudaPackage, CMakePackage, ROCmPackage):
         for arch in CudaPackage.cuda_arch_values:
             if arch not in supported_cuda_arch_list:
                 conflicts("+cuda", when="cuda_arch={0}".format(arch), msg=cuda_msg)
+            depends_on("dla-future cuda_arch={0}".format(arch), when="cuda_arch={0}".format(arch))
+            depends_on("sirius cuda_arch={0}".format(arch), when="cuda_arch={0}".format(arch))
+            depends_on("dbcsr cuda_arch={0}".format(arch), when="cuda_arch={0}".format(arch))
+            depends_on("py-torch cuda_arch={0}".format(arch), when="cuda_arch={0}".format(arch))
 
     with when("+rocm"):
         for arch in ROCmPackage.amdgpu_targets:
             if arch not in supported_rocm_arch_list:
                 conflicts("+rocm", when="amdgpu_target={0}".format(arch), msg=rocm_msg)
+            depends_on(
+                "dla-future amdgpu_target={0}".format(arch), when="amdgpu_target={0}".format(arch)
+            )
+            depends_on(
+                "sirius amdgpu_target={0}".format(arch), when="amdgpu_target={0}".format(arch)
+            )
+            depends_on(
+                "dbcsr amdgpu_target={0}".format(arch), when="amdgpu_target={0}".format(arch)
+            )
+            depends_on(
+                "py-torch amdgpu_target={0}".format(arch), when="amdgpu_target={0}".format(arch)
+            )
```
I think the dependencies need to be qualified with `when="... +dep` as well, otherwise they become unconditional dependencies, no? At least `dla-future` needs a `when="+dlaf"`.

mtaillefumier replied:
agreed. It should be fixed now

msimberg commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -987,7 +1025,7 @@ def cmake_args(self):
             self.define_from_variant("CP2K_USE_VORI", "libvori"),
             self.define_from_variant("CP2K_USE_SPLA", "spla"),
-            self.define_from_variant("CP2K_USE_QUIP", "quip"),
+            # self.define_from_variant("CP2K_USE_QUIP", "quip"),
```
Remove?

mtaillefumier replied:
done

yizeyi18 commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -216,7 +216,7 @@ class Cp2k(MakefilePackage, CudaPackage, CMakePackage, ROCmPackage):
 
     with when("+cosma"):
-        depends_on("cosma+scalapack")
+        depends_on("cosma+scalapack+shared")
```
Without shared library, cp2k +cosma could still get built through makefile method. I suggest using +shared only with "build_system=cmake", or patching cp2k CMakeLists.txt .
cp2k itself do not rely on any cosma shared library. Using +shared only provides "cosma::cosma_pxgemm" target demanded by cp2k CMakeLists.txt. In fact cp2k does not link against this target; cp2k links against libcosma_prefixed_pxgemm and libcosma, provided by both static and shared version.

yizeyi18 commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -109,7 +109,7 @@ class Cp2k(MakefilePackage, CudaPackage, CMakePackage, ROCmPackage):
         " are enabled",
     )
     variant("pytorch", default=False, description="Enable libtorch support")
```
As install with pytorch only implemented with cmake, I suggest add a "when=build_system=cmake" here.

mtaillefumier replied:
+shared is enforced for both build systems. that the best solution

Commit:
removed deprecated versions. additional cleanups

yizeyi18 commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -580,46 +608,20 @@ def edit(self, spec, prefix):
-                fcflags += pkgconf("--cflags", "libint2", output=str).split()
-                libs += pkgconf("--libs", "libint2", output=str).split()
+        fcflags += pkgconf("--cflags", "libint2", output=str).split()
+        libs += pkgconf("--libs", "libint2", output=str).split()
```
Is the indentation correct here? Currently it is out of `if "+libint" in spec:`.

yizeyi18 commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -580,46 +608,20 @@ def edit(self, spec, prefix):
+        # some Fortran functions seem to be direct wrappers of the
+        # C functions such that we get a direct dependency on them,
+        # requiring `-lxc` to be present in addition to `-lxcf03`
+        libs += pkgconf("--libs", "libxcf03", "libxc", output=str).split()
```
Also the indentation here.

mtaillefumier commented:
I am closing this PR because many of the work is done already.
